File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/interpreter/lab/binary_minus.cpp
// binary_minus.cpp

#include <iostream>
#include "binary.h"
#include "binary_minus.h"
#include "dp_exception.h"
class expression;

binary_minus::binary_minus(expression * a_operand1, expression * a_operand2)
: binary(a_operand1, a_operand2)
{
}

void binary_minus::output(std::ostream & a_ostream) const
{
	output_infix(a_ostream, "-");
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/interpreter/solution/BinaryMinus.java
// BinaryMinus.java

import java.io.PrintWriter;

public class BinaryMinus extends Binary
{
	public BinaryMinus(Expression aOperand1, Expression aOperand2)
	{
		super(aOperand1, aOperand2);
	}

	public double getValue(DpMap<String, Double> aMap)
	{
		if (getOperand1() == null)
		{
			throw new DpException("double BinaryMinus.getValue(DpMap<String, Double> aMap) -- getOperand1() == null.");
		}
		double aValue1 = getOperand1().getValue(aMap);

		if (getOperand2() == null)
		{
			throw new DpException("double BinaryMinus.getValue(DpMap<String, Double> aMap) -- getOperand2() == null.");
		}
		double aValue2 = getOperand2().getValue(aMap);

		double result = aValue1 - aValue2;

		return result;
	}

	public void output(PrintWriter aPrintWriter)
	{
		outputInfix(aPrintWriter, "-");
	}
}
